pkg/controller/core.oam.dev/v1alpha2/application/apply.go

```diff
@@ -112,32 +105,14 @@ func (h *appHandler) apply(ctx context.Context, ac *v1alpha2.ApplicationConfigur
 		}
 	}
 	// set the annotation on ac to point out which component are newly changed
-	ac.SetAnnotations(oamutil.MergeMapOverrideWithDst(ac.GetAnnotations(), map[string]string{
+	// make sure that it won't remove the value on the subsequent reconcile loops
+	ac.SetAnnotations(oamutil.MergeMapOverrideWithDst(map[string]string{
 		oam.AnnotationRollingComponent: strings.Join(newComponents, common.RollingComponentsSep),
-	}))
+	}, ac.GetAnnotations()))
 	if err := h.createOrUpdateAppConfig(ctx, ac); err != nil {
 		return err
 	}
 
-	// Garbage Collection for no used Components.
-	// There's no need to ApplicationConfiguration Garbage Collection, it has the same name with Application.
-	for _, comp := range h.app.Status.Components {
-		var exist = false
-		for _, cc := range comps {
-			if comp.Name == cc.Name {
-				exist = true
-				break
-			}
-		}
-		if exist {
-			continue
-		}
-		// Component not exits in current Application, should be deleted
-		var oldC = &v1alpha2.Component{ObjectMeta: metav1.ObjectMeta{Name: comp.Name, Namespace: ac.Namespace}}
-		if err := h.r.Delete(ctx, oldC); err != nil {
-			return err
-		}
-	}
```

Comment:
why delete?

Reply:
no need to garbage collect any components as they all have owner reference to the application. They will be gced automatically.  GC here will make the old AC not working (comp1-> comp2 change then comp1 is deleted which leaves ac-v1 hanging)